src/Routes/Abstract_Route.php

```diff
@@ -0,0 +1,356 @@
+	 *
+	 * @return void
+	 */
+	public function authorize() : void {
+		$this->authorize_request_method();
+	}
+
+	/**
+	 * Override to change the $wp global object.
+	 *
+	 * @param \WP $wp The WP object.
+	 * @return void
+	 */
+	public function parse( \WP $wp ) : void {
+		return;
+	}
+
+	/**
+	 * Override to change the wp_query. This is equivalent to using 'pre_get_posts'.
+	 *
+	 * @param \WP_Query $wp_query The WP query object.
+	 * @return void
+	 */
+	public function before( \WP_Query $wp_query ) : void {
+		return;
+	}
+
+	/**
+	 * Has the template included been fired.
+	 *
+	 * @param string $template The template name.
+	 * @return string          The modified template path.
+	 */
+	public function did_template_include( $template ) {
+		$template_path = $this->get_template();
+
+		// Bail early if no template path.
+		if ( empty( $template_path ) ) {
+			return $template;
+		}
+
+		global $wp_query;
+
+		// Bail early if this is the 404 template.
+		if ( strpos( $template_path, '404' ) !== false ) {
+			$wp_query->is_404 = true;
+			$protocol         = filter_input( INPUT_SERVER, 'SERVER_PROTOCOL', FILTER_SANITIZE_SPECIAL_CHARS );
+			header( "{$protocol} 404 Not Found", true, 404 );
+		}
+
+		// Use the theme path if the passed in path doesn't exist.
+		if ( ! file_exists( $template_path ) ) {
+			$template_path = get_template_directory() . '/' . $template_path;
+		}
+
+		// Bail ealy if the template file exists.
+		if ( file_exists( $template_path ) ) {
+			$wp_query->is_home = false;
+			$this->before( $wp_query );
+			return $template_path;
+		}
+
+		return $template;
+	}
+
+	/**
+	 * Get document title.
+	 *
+	 * @param string $title The current document title.
+	 * @return string       The modified document title.
+	 */
+	public function did_pre_get_document_title( $title ) : string {
+		$route_title = $this->get_title();
+
+		if ( ! empty( $route_title ) ) {
+			$title = $route_title;
+		}
+
+		return $title;
+	}
+
+	/**
+	 * Determines if the current request method is valid for a given route.
+	 *
+	 * @return bool True if the request method is valid, false otherwise.
+	 */
+	public function is_valid_request_method() : bool {
+		$method = filter_input( INPUT_SERVER, 'REQUEST_METHOD', FILTER_SANITIZE_SPECIAL_CHARS );
+
+		if ( empty( $method ) && isset( $_SERVER['REQUEST_METHOD'] ) ) {
+			$method = $_SERVER['REQUEST_METHOD'];
+		}
+
+		return in_array( $method, $this->get_request_methods(), true );
+	}
+
+	/**
+	 * Determine if the request is valid.
+	 *
+	 * @return bool|void True if the request method is valid, throws a 403 forbidden otherwise.
+	 */
+	public function authorize_request_method() {
+		$valid = $this->is_valid_request_method();
+
+		// Bail early if this is the PHPUnit runner.
+		if ( defined( 'PHPUNIT_RUNNER' ) ) {
+			return $valid;
+		}
+
+		// Bail early if the method is valid.
+		if ( $valid ) {
+			return $valid;
+		}
+
+		header( 'HTTP/1.1 403 Forbidden' );
+		wp_die( 'Not Authorized' );
```

Comment:
```suggestion
		wp_die( 'Not Authorized', 403 );
```
Wouldn't this work properly instead?